Replace webmachine with cowboy

stroop23 opened the pull request.

stroop23 committed:
CHANGE replace webmachine with cowboy

stroop23 committed:
CHANGE refactored add_service / remove_service
CHANGE moved function export check and code loading to add_service()
CHANGE removed piqi_rpc_monitor

stroop23 committed:
FIX missing rpc resource.

stroop23 committed:
FIX pick input format in case accept header contains */*

thijsterlouw commented:
First thing I do when so many files / lines changed:  search for "test" and I don't see any changed/new tests, which sounds suspicious :)

The change itself looks pretty good. Even though many lines changed, it is pretty isolated and easy to understand.

thijsterlouw commented on src/piqi_rpc.erl:
```diff
@@ -62,60 +60,52 @@ ensure_started(App) ->
-        ServiceDefs = get_service_defs(),
-        ok = set_service_defs([RpcService | ServiceDefs])
+        ok = check_modules(RpcService),
+        ok = add_service_def(RpcService),
```
add_service_def but not remove_service_def (instead remove_service) , small inconsistency. remove_service internally calls remove_service_def.

stroop23 replied:
? 

add_service does add_service_def, same as remove_service does remove_service_def ?

thijsterlouw commented on src/piqi_rpc_http.erl:
```diff
@@ -21,77 +21,74 @@
+        "/" -> UrlPath;
+        _ -> "/" ++ UrlPath
+    end,
+    AdjustedPath = case string:right(LAdjustedPath, 1) of
```
is this behaviour (of fixing the path by pre/post-pending slashes) consistent with the old behaviour ?

stroop23 replied:
i believe so, i think the tests failed without this. But i'll double check.

thijsterlouw commented on src/piqi_rpc_resource.erl:
```diff
@@ -0,0 +1,195 @@
+rpc(Req, State = #rpc_state{rpc_mod = RpcMod, impl_mod =ImplMod, options = Options}, InputFormat, OutputFormat, Body) ->
+    {Function, Req2} = cowboy_req:binding(function, Req),
+    % Pass the whole request through the process dictionary
+    % @TODO based on options pass it or parts as second argument to the implementation function call.
```
is this TODO important ?

stroop23 replied:
Yes, however the phrasing is wrong. It should state that we need to find a better way to pass request related parameters to the implementation modules.

thijsterlouw commented:
Besides that, LGTM, nice one!

yfyf commented:
It's hard to compare the `piqi_rpc_resource` due to the remove / add. Maybe you could rebase so that there's an actual diff?

Otherwise seems ok, but definitely needs at least some basic tests.

stroop23 committed:
FIX catch unknown content types
FIX more descriptive errors when functions are not implemented in the implementation module.
